lib/share_progress/variation.rb

```diff
@@ -0,0 +1,92 @@
+require 'share_progress'
+require 'share_progress/client'
+
+module ShareProgress
+  class Variation
+    class << self
+      def facebook_type_name
+        'facebook'
+      end
+
+      def twitter_type_name
+        'twitter'
+      end
+
+      def email_type_name
+        'email'
+      end
+
+      def facebook_type
+        FacebookVariation
+      end
+
+      def twitter_type
+        TwitterVariation
+      end
+
+      def email_type
+        EmailVariation
+      end
+    end
+
+    def to_s
+      self.compile_to_hash.to_json
+    end
+
+    def compile_to_hash
+      # override this in variation subclasses
+    end
+
+    def save
+      # this should bump off to the button that controls it to save, just haven't finished it yet
+    end
+
+    def set_analytics(analytics:)
+      # this will be called by the parent button which is able to update the analytics for the variation.
+    end
+  end
+
+  class FacebookVariation < Variation
```

Comment:
Makes sense!
On Aug 25, 2015 4:41 PM, "NealJMD" <notifications@github.com> wrote:

> In lib/share_progress/variation.rb
> <https://github.com/SumOfUs/share_progress/pull/9#discussion_r37916678>:
>
> > +    end
> > +
> > +    def compile_to_hash
> > +      # override this in variation subclasses
> > +    end
> > +
> > +    def save
> > +      # this should bump off to the button that controls it to save, just haven't finished it yet
> > +    end
> > +
> > +    def set_analytics(analytics:)
> > +      # this will be called by the parent button which is able to update the analytics for the variation.
> > +    end
> > +  end
> > +
> > +  class FacebookVariation < Variation
>
> Since the API has different fields for each type of variation, I dig
> having a separate class for each one. Can they go in their own files though?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SumOfUs/share_progress/pull/9/files#r37916678>.
>